### PR TITLE
Invites: Add error tooltip to invalid invitees

### PR DESCRIPTION
**Pull request by lezama**

### CODE REVIEW

**Comment by mtias** [comment]
Should we make the Tooltip red if you pass an `error` prop?

**Comment by rickybanister** [comment]
Yes, we should definitely accommodate 'statuses' for tooltips. That'll be great!

@lezama can we position that tooltip `bottom`?

**Comment by mtias** [comment]
Ok, you can do `<Tooltip status="error">` after https://github.com/Automattic/wp-calypso/commit/8585fe67ea0701780cc494c3bca0f8381d1d179c.

**Comment by lezama** [comment]
@rickybanister: o7

image

**Comment by rickybanister** [comment]
That rules! Nice work @lezama

**Comment by rickybanister** [comment]
One quick question, will we be combining the status prop for the token and for the tooltip? I think both should be red if there's a problem..

**Comment by lezama** [comment]
> One quick question, will we be combining the status prop for the token and for the tooltip? I think both should be red if there's a problem..

the token pass the status to the tooltip, so yes :+1:

**Comment by rickybanister** [comment]
@lezama you asked earlier what should happen when tooltips overlap each other. Here's my take after taking the branch for a spin (sorry if it's complicated):

When you submit a new token for validation and it fails we turn the token text red and we show a red tooltip. You (currently) have the opportunity to type a new entry and move on. I think we could hide the previous tooltip if a new error tooltip is triggered. Then if you hover the previous invalid token we'd show it's tooltip, but not the latest one. Basically we show only one tooltip at a time. If you hover the first invalid entry, then mouseout, we should again show the current invalid tooltip if no new value has been typed into the input. We can talk further Monday if helpful. 

We also need to shorten the errors. If someone types gibberish we don't know if it was meant to be a username OR an email address. We don't really need to display `WordPress.com username does not exist or invalid email address.`

Maybe we need to differentiate errors based on the presence of special characters. If it's just letters we say `Invalid username` and if there's some attempt to type an email (a @ or a .something) we say `Invalid email address`. The tooltips should stay short, otherwise they could go outside the viewport on narrow screens.

**Comment by ebinnion** [comment]
@lezama – I went ahead and rebased this against master. When you get a chance, will you look over it to make sure I didn't break anything accidentally? :stuck_out_tongue_closed_eyes:

**Comment by ebinnion** [comment]
Now that we've added `Tooltip` as a require in `token-field/token.jsx`. It seems that one of `Tooltip`'s dependencies is an npm package called `component-tip`, which uses `require( 'css' )`.

@aduth suggested that we could use mockery to help with this, but I'm having a heck of a time trying to get mockery to work :disappointed:

I noticed that @ehg fixed the `themes-list` tests with mockery. Would you mind lending a hand here?

**Comment by ebinnion** [comment]
Thanks @ehg! Making that change, as well as moving the `TokenFieldWrapper` component out of the same directory as the test, seems to have done the trick.

**Comment by ebinnion** [comment]
@rickybanister – I think the error notices are starting to feel pretty good at this point. I do agree that the tooltips are a bit wordy, but I'd prefer to address those in a separate and more targeted PR.

File: client/my-sites/people/invite-people/index.jsx [hunk 1]
```diff
@@ -69,7 +70,10 @@ export default React.createClass( {
 			return value;
 		} );
 
-		this.setState( { usernamesOrEmails: filteredTokens } );
+		this.setState( {
+			usernamesOrEmails: filteredTokens,
+			errorTokenHover: ! includes( filteredTokens, errorTokenHover ) ? false : errorTokenHover
```
Comment (lezama) [inline client/my-sites/people/invite-people/index.jsx]: I think we can write this like: `errorTokenHover: includes( filteredTokens, errorTokenHover ) && errorTokenHover`

File: client/components/token-field/test/index.jsx [hunk 2]
```diff
@@ -92,6 +98,14 @@ describe( 'TokenField', function() {
...
 
 	beforeEach( function() {
+		var TokenFieldWrapper;
+		mockery.registerMock( 'components/tooltip', EMPTY_COMPONENT );
+		mockery.enable( {
+			warnOnReplace: false,
+			warnOnUnregistered: false
+		} );
+
+		TokenFieldWrapper = require( './lib/token-field-wrapper' );
```
Comment (lezama) [inline client/components/token-field/test/index.jsx]: We could use `let` here.

File: client/components/token-field/test/index.jsx [hunk 3]
```diff
@@ -92,6 +98,14 @@ describe( 'TokenField', function() {
 	} );
 
 	beforeEach( function() {
+		var TokenFieldWrapper;
+		mockery.registerMock( 'components/tooltip', EMPTY_COMPONENT );
```
Comment (lezama) [inline client/components/token-field/test/index.jsx]: I think we can register this on `before` instead of `beforeEach`, there's an `after` hook too where we can put the deregistration.

File: client/my-sites/people/invite-people/index.jsx (hunk 1 above)
Comment (ebinnion) [inline client/my-sites/people/invite-people/index.jsx]: We could, although I honestly find that style a bit hard to read at times.

File: client/my-sites/people/invite-people/index.jsx [hunk 4]
```diff
@@ -94,12 +102,24 @@ export default React.createClass( {
...
+			let status, tooltip;
 			if ( errors && errors[ value ] ) {
 				status = 'error';
+
+				// We only want to show one tooltip.
+				if ( ! errorTooltipAdded ) {
+					if ( ! errorTokenHover || ( errorTokenHover && value === errorTokenHover ) ) {
+						// Attempt to get the error message for the tooltip, and set errorTooltipAdded
+						// to true if message was found.
+						let errorValue = errorTokenHover ? errorTokenHover : value;
```
Comment (lezama) [inline client/my-sites/people/invite-people/index.jsx]: `errorValue` could be a `const` :stuck_out_tongue_closed_eyes:

Comment (lezama) [inline client/my-sites/people/invite-people/index.jsx]: we could write this `const errorValue = errorTokenHover || value;

File: client/my-sites/people/invite-people/index.jsx [hunk 5]
```diff
@@ -107,7 +127,10 @@ export default React.createClass( {
 			if ( status ) {
 				value = {
 					value,
-					status
+					status,
+					tooltip,
+					onMouseEnter: () => this.setState( { errorTokenHover: value.value } ),
+					onMouseLeave: this.onErrorTokenMouseLeave
```
Comment (lezama) [inline client/my-sites/people/invite-people/index.jsx]: I think we could put both `onMouseEnter` and `onMouseEnter` inline.
